pyro/infer/abstract_infer.py

```diff
@@ -87,6 +89,44 @@ def run(self, *args, **kwargs):
         self._categorical = Categorical(logits=torch.tensor(self.log_weights))
         return self
 
+    def information_criterion(self, pointwise=False):
+        """
+        Computes information criterion of the model. Currently, returns only "Widely
+        Applicable/Watanabe-Akaike Information Criterion" (WAIC) and the corresponding
+        effective number of parameters.
+
+        Reference:
+
+        [1] `Practical Bayesian model evaluation using leave-one-out cross-validation and WAIC`,
+        Aki Vehtari, Andrew Gelman, and Jonah Gabry
+
+        :param bool pointwise: a flag to decide if we want to get a vectorized WAIC or not. When
+            ``pointwise=False``, returns the sum.
+        :returns OrderedDict: a dictionary containing values of WAIC and its effective number of
+            parameters.
+        """
+        if not self.exec_traces:
+            return {}
+        obs_node = None
+        log_likelihoods = []
+        for trace in self.exec_traces:
+            obs_nodes = trace.observation_nodes
+            if len(obs_nodes) > 1:
+                raise ValueError("Infomation criterion calculation only works for models "
```

Comment:
It seems that we should be able to flatten all our observations and support `len(obs_nodes) > 1`, right? Not saying we should do that, but just want to understand where this restriction is coming from.

Reply:
@neerajprad I can't find any documentation for that but your suggestion is quite reasonable! Will make a change if you (or other users in the future) need it.

Reply:
Ah, indeed I don't know what we should do with the case of many observations. In its theory, each sample gives a likelihood. And waic is calculated from a batch of samples (with a batch of likelihoods). If we allow many observations, I am not sure that we should concatenate their likelihoods, or sum them, or take average of them. >"<